### PR TITLE
Add telemetry QOS

### PR DESCRIPTION
**Pull request by danewalton**
closes #109

### CODE REVIEW

**Comment by danewalton** [comment]
https://github.com/Azure/azure-iot-middleware-freertos/pull/109/checks?check_run_id=2496660158#step:5:1035

Lines `1035-1050` for logs of waiting for puback.

File: source/azure_iot_hub_client.c [hunk 1]
```diff
@@ -126,6 +137,32 @@ static void prvMQTTProcessSuback( AzureIoTHubClient_t * pxAzureIoTHubClient,
...
+{
+    ( void ) pxIncomingPacket;
+
+    configASSERT( pxIncomingPacket != NULL );
+    configASSERT( ( azureiotmqttGET_PACKET_TYPE( pxIncomingPacket->ucType ) ) == azureiotmqttPACKET_TYPE_PUBACK );
+
+    if( pxAzureIoTHubClient->_internal.xTelemetryAckContext._internal.usMqttPubPacketID == usPacketId )
+    {
+        pxAzureIoTHubClient->_internal.xTelemetryAckContext._internal.usState = azureiothubTELEMETRY_ACK_STATE_PUBACK;
+        AZLogInfo( ("Puback received for packet id: 0x%08x", usPacketId) );
```
Comment (hihigupt) [inline source/azure_iot_hub_client.c:157]: nit: ( ( " " ) );

File: source/azure_iot_hub_client.c [hunk 2]
```diff
@@ -831,7 +924,15 @@ AzureIoTHubClientResult_t AzureIoTHubClient_SendTelemetry( AzureIoTHubClient_t *
             AZLogError( ( "Failed to publish telemetry: MQTT error=0x%08x", xMQTTResult ) );
             xResult = eAzureIoTHubClientPublishFailed;
         }
-        else
+
+        if( ( xMQTTResult == eAzureIoTMQTTSuccess ) && ( xQOS == eAzureIoTHubMessageQoS1 ) )
```
Comment (hihigupt) [inline source/azure_iot_hub_client.c:928]: else if ( ( xQOS == eAzureIoTHubMessageQoS1 ) && (xResult =prvWaitForPubAck() ) != eAzureIoTHubClientSuccess )
{
Log_Error("Failed to wait for pubACK")
}
else
{
   AZLogInfo( ( "Successfully sent telemetry message" ) );
   xResult = eAzureIoTHubClientSuccess;
}

File: source/azure_iot_hub_client.c [hunk 3]
```diff
@@ -462,6 +503,56 @@ static AzureIoTHubClientResult_t prvWaitForSubAck( AzureIoTHubClient_t * pxAzure
...
+            ulWaitTime = ulTimeoutMilliseconds;
+            ulTimeoutMilliseconds = 0;
+        }
+
+        if( AzureIoTMQTT_ProcessLoop( &pxAzureIoTHubClient->_internal.xMQTTContext, ulWaitTime ) != eAzureIoTMQTTSuccess )
+        {
+            xResult = eAzureIoTHubClientFailed;
+            break;
+        }
+    } while( ulTimeoutMilliseconds );
```
Comment (hihigupt) [inline source/azure_iot_hub_client.c:544]: is it possible to merge this code with above code?

File: source/include/azure_iot_hub_client.h [hunk 4]
```diff
@@ -165,6 +171,15 @@ typedef struct AzureIoTHubClientReceiveContext
     } _internal;
 } AzureIoTHubClientReceiveContext_t;
 
+typedef struct AzureIoTHubClientTelemetryAckContext
```
Comment (hihigupt) [inline source/include/azure_iot_hub_client.h:174]: >AzureIoTHubClientTelemetryAckContext [](http://example.com/codeflow?start=15&length=36)

Should we name it publish ack context?

File: tests/e2e/device/e2e_device_commands.c [hunk 5]
```diff
@@ -962,7 +963,7 @@ uint32_t ulE2EDeviceProcessCommands( AzureIoTHubClient_t * pxAzureIoTHubClient )
     if( AzureIoTHubClient_SendTelemetry( pxAzureIoTHubClient,
                                          e2etestE2E_TEST_CONNECTED_MESSAGE,
                                          sizeof( e2etestE2E_TEST_CONNECTED_MESSAGE ) - 1,
-                                         NULL ) != eAzureIoTHubClientSuccess )
+                                         NULL, eAzureIoTHubMessageQoS1, 10 ) != eAzureIoTHubClientSuccess )
```
Comment (hihigupt) [inline tests/e2e/device/e2e_device_commands.c:966]: >10 [](http://example.com/codeflow?start=72&length=2)

should we use WAIT_FOREVER(ULONG_MAX)

File: source/include/azure_iot_hub_client.h (hunk 4 above)
Comment (danewalton) [reply source/include/azure_iot_hub_client.h:174]: Yup that way we allow for other features should they arise.

File: source/azure_iot_hub_client.c (hunk 3 above)
Comment (danewalton) [reply source/azure_iot_hub_client.c]: please take a look at the most recent changes. I've merged the `waitForX` stuff into one and it means consolidation of other things like here: https://github.com/Azure/azure-iot-middleware-freertos/blob/7778f751e10309f4ff6e89f499d27f677ff61521/source/azure_iot_hub_client.c#L36-L41

and here:
https://github.com/Azure/azure-iot-middleware-freertos/blob/7778f751e10309f4ff6e89f499d27f677ff61521/source/include/azure_iot_hub_client.h#L54

**Comment by danewalton** [comment]
closing in favor of #113

**Pull request closed without merging**